### ISSUE TITLE
TaskExecutionResult specify value when property changes

### ISSUE THREAD

**Issue by mathjeff** [gradle/gradle#18045]
### Expected Behavior
When an OperationCompletionListener receives a TaskExecutionResult, it would be nice if the the executionReasons specified not only which properties had changed but what their values had changed from and to.

### Current Behavior
Currently when OperationCompletionListener receives a TaskExecutionResult, the reasons may say:

```
[Value of input property 'totalMemory' has changed for task ':saveSystemStats']
```

It would be nice if this could look more like:

```
[Value of input property 'totalMemory' has changed for task ':saveSystemStats', from "10" to "20"]
```

### Context
AndroidX has a check to run the build twice in a row and ensure that the second run is all UP-TO-DATE , see https://github.com/gradle/gradle/issues/13753#issuecomment-846219711

Occasionally this check fails  ( https://issuetracker.google.com/196115864 , failing task at https://android.googlesource.com/platform/frameworks/support/+/b94097ad782a795d6b42fadc1822ae633dea6b60/buildSrc/private/src/main/kotlin/androidx/build/SaveSystemStatsTask.kt ) and it would be helpful to know what the value changed from and what it changed to, to better understand what is wrong.

Thanks!

**Comment by mathjeff** [comment]
Based on looking at the code, it seems that

https://github.com/gradle/gradle/blob/68e638a775b01e5bbd22735aaeff490adf029370/subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/InputValueChanges.java#L54

might be where this string gets created

**Comment by mlopatkin** [comment]
@gradle/execution could you please take a look?

**Comment by wolfs** [comment]
Hey @mathjeff. Yes, Gradle could do this for certain types of properties, like `String`s or `Enum`s. There is an inherent security risk you declare secrets as inputs. You shouldn't do that, though nobody stops you from doing so. So the question is somewhat if you should be able to opt-in or opt-out of the behaviour to show the actual value of a property.

What we currently suggest folks doing is to print out the actual task properties in a `doFirst` task action for the interesting tasks and then compare those things across builds. I agree that this is cumbersome and wonder if we could build a feature like that instead.

**Comment by mathjeff** [comment]
Hey thanks @wolfs , that's interesting.

Does that mean that the concern is that a non-malicious OperationCompletionListener might print a secret value and the user might inadvertently share the resulting build log too widely?

I haven't been able to think of an argument for why that would never happen in Gradle builds in general, but can I add some more details about why AndroidX wouldn't be concerned about it.

I don't that AndroidX is declaring any secrets as task inputs (as expected).

Also, In our particular case we're only printing these values when we're also failing the build, so in our particular case even if secrets did manage to become the values of some input properties, our users would still notice them and make sure not to share them (more discussion about finding the relevant output in a failing build:  https://github.com/gradle/gradle/issues/13090#issuecomment-758800391 ).

For the moment we're essentially doing what you've described involving separately saving the relevant values (in this case, making a backup of the output file) in the specific cases when we know we need them.

**Comment by wolfs** [comment]
> Does that mean that the concern is that a non-malicious OperationCompletionListener might print a secret value and the user might inadvertently share the resulting build log too widely?

Yes. This is especially true for build scans, since we capture the message there as well. For example in our build, we publish build scans to the public ge.gradle.org instance. Although we also don't use secrets as inputs, so it wouldn't be a problem for us.

**Comment by wolfs** [comment]
We don't plan to print out the value in the changed value with the current `@Input` annotation. We could imagine us introducing a new way to say that something isn't a secret, which would allow us to print the changed value.
